src/platform/unix/mod.rs

```diff
@@ -415,80 +417,98 @@ impl OsIpcChannel {
 
 pub struct OsIpcReceiverSet {
     incrementor: Incrementor,
-    pollfds: Vec<pollfd>,
-    fdids: HashMap<c_int, u64, BuildHasherDefault<FnvHasher>>
+    poll: Poll,
+    pollfds: HashMap<Token, c_int, BuildHasherDefault<FnvHasher>>,
+    fdids: HashMap<c_int, u64, BuildHasherDefault<FnvHasher>>,
+    events: Events
 }
 
 impl Drop for OsIpcReceiverSet {
     fn drop(&mut self) {
-        unsafe {
-            for pollfd in self.pollfds.iter() {
-                let result = libc::close(pollfd.fd);
-                assert!(thread::panicking() || result == 0);
+        let mut result = 0;
+        for (_, &fd) in &self.pollfds {
+            unsafe {
+                result |= libc::close(fd);
             }
-        }
+        };
+        assert!(thread::panicking() || result == 0);
     }
 }
 
 impl OsIpcReceiverSet {
     pub fn new() -> Result<OsIpcReceiverSet,UnixError> {
-        let fnv = BuildHasherDefault::<FnvHasher>::default();
+        let fnv1 = BuildHasherDefault::<FnvHasher>::default();
+        let fnv2 = BuildHasherDefault::<FnvHasher>::default();
         Ok(OsIpcReceiverSet {
             incrementor: Incrementor::new(),
-            pollfds: Vec::new(),
-            fdids: HashMap::with_hasher(fnv)
+            poll: Poll::new().expect("Could not create a new receiver"),
```

Comment:
AFAIK on linux this will be an abstraction of epoll_create1, so the possible source of errors will be the following

> EMFILE
The per-user limit on the number of epoll instances imposed by /proc/sys/fs/epoll/max_user_instances was encountered. See epoll(7) for further details.
ENFILE
The system limit on the total number of open files has been reached.
ENOMEM
There was insufficient memory to create the kernel object.